### PR TITLE
Improve UX of TypedData usage

### PR DESCRIPTION
**Pull request by PhilippeR26**
## Motivation and Resolution
Solves #529 
Several users have reported difficulties to use signature and verification of `TypedData` (SNIP-12, EIP712), especially with related  `Account` methods.

- `Account.verifyMessage()` & `Account.verifyMessageHash()` are shadowed.
- A `Provider.verifyMessageInStarknet()` method has been added.
- An `ec.getFullPublikey()` helper has been added.
- A `typedData.verifyMessage()` function has been added.

- As `Account.verifyMessage()` & `Account.verifyMessageHash()` needs to instantiate a fake Account with a fake private key, these functions are shadowed (tagged as @ deprecated) and are replaced by `Provider.verifyMessageInStarknet()`.
  It clarify also that this verification is working by interacting with Starknet.
```typescript
const isValid4 = await myProvider.verifyMessageInStarknet(myTypedData, signature2, account0.address);
const isValid5 = await myProvider.verifyMessageInStarknet(msgHash, signature2, account0.address);
```
- the complicated code to get the full public key 
```typescript
const fullPubKey = encode.addHexPrefix(encode.buf2hex(ec.starkCurve.getPublicKey(privateKey, false)));
```
   can now also be made with 
```typescript
const fullPublicKey = ec.getFullPublicKey(privateKey);
```

- It was problematic to verify a message outside of Starknet, with usage of low level functions and problematic types ( Uint8Array, ArraySignature not supported, ...).

- We have now a dedicated function, with friendly types (`BigNumberish`, `Signature`, ...) :
```typescript
const isValid = typedData.verifyMessage(myTypedData, signature, fullPublicKey, account0Address);
const isValid2 = typedData.verifyMessage(msgHash, signature, fullPublicKey);
```

## Development related changes
- Interfaces: `verifyMessage()` & `verifyMessageHash()` have been removed from the Account interface. 
  `verifyMessageInStarknet()`has not been added in RpcProvider class, to not impact users implementations.
- code of  `Account.verifyMessage()` & `Account.verifyMessageHash()` has been refactored in  `Provider.verifyMessageInStarknet()`.
- Some utils functions have been added : `num.isBigNumberish()`,  `typedData.verifyMessage()` & `ec.getFullPublikey()`
- tests : tests related to new functions have been added.
- docs : All JSDoc and guides for  `Account.verifyMessage()` & `Account.verifyMessageHash()` have been removed. 
  All new functions includes JSDoc, and have been explained in the Signature guide.

## Checklist:

- [x] Performed a self-review of the code
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Linked the issues which this PR resolves
- [x] Documented the changes in code (API docs will be generated automatically)
- [x] Updated the tests
- [x] All tests are passing

### CODE REVIEW

File: __tests__/rpcProvider.test.ts [hunk 1]
```diff
@@ -493,3 +493,49 @@ describeIfNotDevnet('waitForBlock', () => {
     expect(true).toBe(true); // answer without timeout Error (blocks have to be spaced with 16 minutes maximum : 200 retries * 5000ms)
   });
 });
+
+describeIfDevnet('EIP712 verification', () => {
+  // currently only in Devnet-rs, because can fail in Sepolia.
+  // to test in all cases once PR#989 implemented.
```
Comment (ivpavici) [inline __tests__/rpcProvider.test.ts:499]: This comment has been addressed through https://github.com/starknet-io/starknet.js/pull/1128

So, we can either remove it or add more tests

File: www/docs/guides/signature.md [hunk 2]
```diff
@@ -51,8 +48,11 @@ The sender provides the message, the signature, and the full public key. Verific
 
 ```typescript
 const msgHash1 = hash.computeHashOnElements(message);
-const result1 = ec.starkCurve.verify(signature, msgHash1, fullPublicKey);
-console.log('Result (boolean) =', result1);
+const isValid1 = typedData.verifyMessage(msgHash1, signature, fullPublicKey);
+console.log('Result (boolean) =', isValid1);
+
+// with a low level function (take care to Types limitations) :
```
Comment (ivpavici) [inline www/docs/guides/signature.md:54]:
```suggestion
// with a low level function (take care of Types limitations) :
```

File: www/docs/guides/signature.md [hunk 3]
```diff
@@ -73,122 +73,121 @@ Check that the Public Key of the account is part of the full public Key:
...
+```typescript
+const isValid = typedData.verifyMessage(myTypedData, signature, fullPublicKey, account0Address);
+```
+
+A verification is also possible if you have the message hash, the signature and the full public key.
+
+```typescript
+const isValid2 = typedData.verifyMessage(msgHash, signature, fullPublicKey);
+
+// with a low level function (take care to Types limitations) :
```
Comment (ivpavici) [inline www/docs/guides/signature.md:170]:
```suggestion
// with a low level function (take care of Types limitations) :
```

Comment (PhilippeR26) [reply www/docs/guides/signature.md:170]: Implemernted

File: www/docs/guides/signature.md (hunk 2 above)
Comment (PhilippeR26) [reply www/docs/guides/signature.md:54]: Implemented

File: __tests__/rpcProvider.test.ts (hunk 1 above)
Comment (PhilippeR26) [reply __tests__/rpcProvider.test.ts:499]: If the problem is addressed, then it can be made in all networks.
Modified.

Comment (ivpavici) [reply __tests__/rpcProvider.test.ts]: I didn't verify about devnet vs testnet :D I just saw that PR#989 was merged

Comment (PhilippeR26) [reply __tests__/rpcProvider.test.ts]: Well, my PR#989 has not been merged : Tony closed it, implementing its own PR#1128 on this subject.
So, I suppose that he made the necessary tests to validate that it now works everywhere in all networks.

File: src/utils/ec.ts [hunk 4]
```diff
@@ -1,3 +1,23 @@
...
+ * get the hex string of the full public key related to a Starknet private key.
+ * @param {BigNumberish} privateKey a 252 bits private key.
+ * @returns {string} an hex string of a 520 bit number, representing the full public key related to `privateKey`.
+ * @example
+ * ```typescript
+ * const result = ec.getFullPublicKey("0x43b7240d227aa2fb8434350b3321c40ac1b88c7067982549e7609870621b535");
+ * // result = "0x0400b730bd22358612b5a67f8ad52ce80f9e8e893639ade263537e6ef35852e5d3057795f6b090f7c6985ee143f798608a53b3659222c06693c630857a10a92acf"
+ * ```
+ */
+export function getFullPublicKey(privateKey: BigNumberish): string {
```
Comment (penovicp) [inline src/utils/ec.ts:19]: I would prefer to avoid expanding `utils/ec.ts` to keep its identity as a submodule alias. I would suggest moving this method to `utils/stark.ts`.

File: src/utils/ec.ts [hunk 5]
```diff
@@ -1,3 +1,23 @@
...
+  const privKey = addHexPrefix(BigInt(privateKey).toString(16));
```
Comment (penovicp) [inline src/utils/ec.ts:20]:
```suggestion
  const privKey = toHex(privateKey);
```

**Review by penovicp — COMMENTED** [review]
Made a few suggestions for some minor changes, the test filter and `isBigNumberish` edge case being most notable.

In general lgtm.

File: src/utils/typedData.ts [hunk 6]
```diff
@@ -596,3 +598,51 @@ export function getMessageHash(typedData: TypedData, account: BigNumberish): str
...
+ * const result2 = typedData.verifyMessage(messageHash, sign, fullPubK);
+ * // result1 = result2 = true
+ * ```
+ */
+export function verifyMessage(
+  message: BigNumberish | TypedData,
+  signature: Signature,
+  fullPublicKey: BigNumberish,
+  accountAddress?: BigNumberish
+): boolean {
```
[the remark this answers is not in the document]
Comment (PhilippeR26) [reply src/utils/typedData.ts:626]: Implemented.

File: src/utils/num.ts [hunk 7]
```diff
@@ -397,3 +397,22 @@ export function stringToSha256ToArrayBuff4(str: string): Uint8Array {
...
+ * ```typescript
+ * const res = num.isBigNumberish("ZERO");
+ * // res = false
+ *  ```
+ */
+export function isBigNumberish(input: unknown): input is BigNumberish {
+  return (
+    isNumber(input) ||
+    isBigInt(input) ||
+    (typeof input === 'string' && (isHex(input) || /^[0-9]*$/i.test(input)))
```
[the remark this answers is not in the document]
Comment (PhilippeR26) [reply src/utils/num.ts:416]: Implemented.

File: src/utils/ec.ts (hunk 5 above)
Comment (PhilippeR26) [reply src/utils/ec.ts:20]: Not possible. 
`Dependency cycle`

File: src/utils/ec.ts (hunk 4 above)
Comment (PhilippeR26) [reply src/utils/ec.ts:19]: Moved.

File: __tests__/utils/ec.test.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import { ec } from '../../src';
+
+describe.only('ec full public key', () => {
```
[the remark this answers is not in the document]
Comment (PhilippeR26) [reply __tests__/utils/ec.test.ts:3]: Moved in stark test

File: src/utils/typedData.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (PhilippeR26) [reply src/utils/typedData.ts:626]: @penovicp 
Once modified, the code is not compiling :
```
src/utils/typedData.ts:632:17
    632 export function verifyMessage(
                        ~~~~~~~~~~~~~
    The call would have succeeded against this implementation, but implementation signatures of overloads are not externally visible.
```
What's this???

[the remark this answers is not in the document]
Comment (PhilippeR26) [reply src/utils/typedData.ts:626]: Don't ask me why, but this is working :
```typescript
export function verifyMessage(
  message: TypedData,
  signature: Signature,
  fullPublicKey: BigNumberish,
  accountAddress: BigNumberish
): boolean;
export function verifyMessage(
  message: BigNumberish,
  signature: Signature,
  fullPublicKey: BigNumberish
): boolean;
export function verifyMessage(
  message: BigNumberish | TypedData,
  signature: Signature,
  fullPublicKey: BigNumberish,
  accountAddress?: BigNumberish
): boolean;
export function verifyMessage(
  message: BigNumberish | TypedData,
  signature: Signature,
  fullPublicKey: BigNumberish,
  accountAddress?: BigNumberish
): boolean {
```

[the remark this answers is not in the document]
Comment (penovicp) [reply src/utils/typedData.ts:643]: The error is probably from one of the failure tests where the parameters are deliberately wrong. I'll double check soon and add a tweak if I'm right. Either way I'll approve the PR.

**Review by penovicp — APPROVED** [review]
